lib/server-factory.js

```diff
@@ -0,0 +1,129 @@
+const http = require('http')
+const https = require('https')
+const http2 = require('http2')
+const net = require('net')
+const { Transform, PassThrough } = require('stream')
+const tls = require('tls')
+const WebSocket = require('ws')
+
+const defaultOptions = {
+  ws: false,
+  https: false,
+  http2: false,
+  tls: false,
+  trustProxy: false,
+  serverFactory: null,
+  protocolDecoder: protocolDecoder
+}
+
+const createServer = (options, aedesHandler) => {
+  if (!aedesHandler) {
+    throw new Error('Missing aedes handler')
+  }
+
+  options = Object.assign({}, defaultOptions, options)
+
+  let server = null
+  if (options.serverFactory) {
+    server = options.serverFactory(aedesHandler, options)
+  } else if (options.tls) {
+    server = tls.createServer(options, aedesHandler)
+  } else if (options.ws) {
+    if (options.https) {
+      if (options.http2) {
+        server = http2.createSecureServer(options.https)
+      } else {
+        server = https.createServer(options.https)
+      }
+    } else {
+      if (options.http2) {
+        server = http2.createServer()
+        // server.on('session', sessionTimeout(options.http2SessionTimeout))
+      } else {
+        server = http.createServer()
+      }
+    }
+    const ws = new WebSocket.Server({ server })
+    ws.on('connection', (conn, req) => {
+      const stream = WebSocket.createWebSocketStream(conn)
+      stream._socket = conn._socket
+      bindConnection(options, aedesHandler, stream, req)
+    })
+  } else {
+    server = net.createServer((conn) => {
+      bindConnection(options, aedesHandler, conn)
+    })
+  }
+  return server
+}
+
+const bindConnection = (options, aedesHandler, conn, req) => {
+  if (options.trustProxy) {
+    extractConnectionDetails(options, aedesHandler, conn, req)
+  } else {
+    aedesHandler(conn, req)
+  }
+}
+
+const extractConnectionDetails = (options, aedesHandler, conn, req = {}) => {
+  const rewindableStream = conn.pipe(new Rewindable())
+
+  conn.on('data', (buffer) => {
+    // TODO : requires change in protocolDecoder to remove the client arg
+    const broker = { trustProxy: options.trustProxy }
+    const decoderArgs = { broker, conn, req }
+    const protocol = options.protocolDecoder(decoderArgs, buffer)
+
+    let stream = null
+    if (buffer && protocol.data) {
+      const removeProxyHeader = removeFromStream(
+        buffer.length - protocol.data.length
+      )
+      stream = rewindableStream.rewind().pipe(removeProxyHeader)
+    } else {
+      stream = rewindableStream.rewind()
+    }
+    req.connDetails = protocol
+    aedesHandler(stream, req)
+  })
+}
+
+class Rewindable extends Transform {
+  constructor () {
+    super()
+    this.accumulator = []
+  }
+
+  _transform (chunk, encoding, callback) {
+    this.accumulator.push(chunk)
+    callback()
+  }
+
+  rewind () {
+    const stream = new PassThrough()
+    this.accumulator.forEach((chunk) => stream.write(chunk))
+    return stream
+  }
+}
```

Comment:
True, even though it only happens when a new client connects to the server.
I'd like to avoid, but i don't see how to achieve the following :
- read the stream to extract/parse the PROXY header
- go back to the initial state of the stream (rewind)
- remove the PROXY header from the stream so aedes only receives raw MQTT frames

I'm not really a pro in stream handling so i would be happy to get some feedback on how to achieve this in a better way.